flytepropeller/pkg/controller/nodes/common/utils.go

```diff
@@ -67,3 +78,82 @@ func GetTargetEntity(ctx context.Context, nCtx interfaces.NodeExecutionContext)
+	if inferredType == nil {
+		errString := "Failed to determine literal type for offloaded literal"
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+
+	// offload the literal
+	if err := datastore.WriteProtobuf(ctx, dataReference, storage.Options{}, toBeOffloaded); err != nil {
+		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
+		return err
+	}
+
+	// update the literal with the offloaded URI, size and inferred type
+	toBeOffloaded.Value = &idlcore.Literal_OffloadedMetadata{
+		OffloadedMetadata: &idlcore.LiteralOffloadedMetadata{
+			Uri:          dataReference.String(),
+			SizeBytes:    literalSizeBytes,
+			InferredType: inferredType,
+		},
+	}
+	logger.Infof(ctx, "Offloaded literal at location [%s] with size [%d] MB and inferred type [%s]", dataReference, literalSizeMB, inferredType)
+	return nil
+}
+
+// CheckOffloadingCompat checks if the upstream and downstream nodes are compatible with the literal offloading feature and returns an error if not contained in phase info object
+func CheckOffloadingCompat(ctx context.Context, nCtx interfaces.NodeExecutionContext, inputLiterals map[string]*core.Literal, node v1alpha1.ExecutableNode, literalOffloadingConfig config.LiteralOffloadingConfig) *handler.PhaseInfo {
+	consumesOffloadLiteral := false
+	for _, val := range inputLiterals {
+		if val != nil && val.GetOffloadedMetadata() != nil {
+			consumesOffloadLiteral = true
+			break
+		}
+	}
+	var phaseInfo handler.PhaseInfo
+	switch node.GetKind() {
+	case v1alpha1.NodeKindTask:
+		taskID := *node.GetTaskID()
+		taskNode, err := nCtx.ExecutionContext().GetTask(taskID)
+		if err != nil {
+			phaseInfo = handler.PhaseInfoFailure(core.ExecutionError_SYSTEM, "GetTaskIDFailure", err.Error(), nil)
+			return &phaseInfo
+		}
+		runtimeData := taskNode.CoreTask().GetMetadata().GetRuntime()
+		if consumesOffloadLiteral && !literalOffloadingConfig.IsSupportedSDKVersion(runtimeData.GetType().String(), runtimeData.GetVersion()) {
+			if !literalOffloadingConfig.Enabled {
+				errMsg := fmt.Sprintf("task [%s] is trying to consume offloaded literals but feature is not enabled", taskID)
+				logger.Errorf(ctx, errMsg)
+				phaseInfo = handler.PhaseInfoFailure(core.ExecutionError_USER, "LiteralOffloadingDisabled", errMsg, nil)
+				return &phaseInfo
+			}
+			leastSupportedVersion := literalOffloadingConfig.GetSupportedSDKVersion(runtimeData.GetType().String())
+			errMsg := fmt.Sprintf("Literal offloading is not supported for this task as its registered with SDK version [%s] which is less than the least supported version [%s] for this feature", runtimeData.GetVersion(), leastSupportedVersion)
+			logger.Errorf(ctx, errMsg)
+			phaseInfo = handler.PhaseInfoFailure(core.ExecutionError_USER, "LiteralOffloadingNotSupported", errMsg, nil)
+			return &phaseInfo
+		}
+	default:
+		logger.Warnf(ctx, "literal offloading : skipping sdk version check for node kind '%s'", node.GetKind())
```

Comment:
should we log the node id here too?

Reply:
cc @hamersaw is there any way to validate the SDK version for downstream array nodes here?